### PR TITLE
SW-2392 module Interlock switches I/O DTS

### PR DESCRIPTION
**Pull request by amrithvenkat**

### CODE REVIEW

File: drivers/gpio/gpio_pcf8575.c [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+ *
+ * @param dev Pointer to the device structure of the driver instance.
+ * @param value Pointer to the input value. It contains the received Bytes(receives 2 Bytes for P0 and P1).
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error code.
+ */
+static int pcf8575_process_input(const struct device *dev, gpio_port_value_t *value)
+{
+	const struct pcf8575_drv_cfg *drv_cfg = dev->config;
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c:60]: General pointer usage rule. Never use a pointer that is passed to a function as an argument without checking first if it is null or not.

File: drivers/gpio/gpio_pcf8575.c [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+	struct pcf8575_drv_data *drv_data = dev->data;
+	int rc = 0;
+	uint16_t rx_buf;
+	uint8_t rx_buf_helper[2];
+	
+	rc = i2c_read_dt(&drv_cfg->i2c, rx_buf_helper, sizeof(rx_buf_helper));
+	if (rc != 0) {
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c]:
```suggestion
	if (rc) {
```

File: drivers/gpio/gpio_pcf8575.c [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+		LOG_ERR("%s: failed to read from device: %d", dev->name, rc);
+		return -EIO;
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c:75]: what is "-EIO" ?

File: drivers/gpio/gpio_pcf8575.c [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+ * @retval 0 If successful.
+ * @retval Negative value for error code.
+ */
+static int pcf8575_process_input(const struct device *dev, gpio_port_value_t *value)
+{
+	const struct pcf8575_drv_cfg *drv_cfg = dev->config;
+	struct pcf8575_drv_data *drv_data = dev->data;
+	int rc = 0;
+	uint16_t rx_buf;
+	uint8_t rx_buf_helper[2];
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c]: Magic number 2 , Create a define for it

File: drivers/gpio/gpio_pcf8575.c [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+	
+	rc = i2c_read_dt(&drv_cfg->i2c, rx_buf_helper, sizeof(rx_buf_helper));
+	if (rc != 0) {
+		LOG_ERR("%s: failed to read from device: %d", dev->name, rc);
+		return -EIO;
+	}
+	rx_buf = rx_buf_helper[0];		// P07-P00 first byte received(read) from pcf8575
+	rx_buf |= rx_buf_helper[1]<<8;	// P17-P07 second byte received(read) from pcf8575 whose value is shifted by 8 and stored in *value
+	
+	if (value) {
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c]: what if value is null ?

File: drivers/gpio/gpio_pcf8575.c [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+		*value = rx_buf; // format P17-P10..P07-P00 (bit15-bit8..bit7-bit0)
+	}
+
+	drv_data->input_port_last = rx_buf; 
+
+	return rc;
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c:91]: so here if we reach this point and the value pointer is Null, we will return 0 as well (Success), is this intended ?

File: drivers/gpio/gpio_pcf8575.c [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+
+	uint32_t changed_pins;
+	uint16_t input_port_last_temp = drv_data->input_port_last;
+	int rc = pcf8575_process_input(drv_data->dev, &changed_pins);
+
+	if (rc) {
+		LOG_ERR("Failed to read interrupt sources: %d", rc);
+	}
+	k_sem_give(&drv_data->lock);
+	if (input_port_last_temp != (uint16_t)changed_pins && !rc) {
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c]: add more Parenthesis so that it is more readable

File: drivers/gpio/gpio_pcf8575.c [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+ * @retval Negative value for error code.
+ */
+static int pcf8575_port_set_raw(const struct device *dev, uint16_t mask, uint16_t value,
+				uint16_t toggle)
+{
+	const struct pcf8575_drv_cfg *drv_cfg = dev->config;
+	struct pcf8575_drv_data *drv_data = dev->data;
+	int rc = 0;
+	uint16_t tx_buf;
+	uint8_t tx_buf_p[2];
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c]: same magic number

File: drivers/gpio/gpio_pcf8575.c [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+		return -EOPNOTSUPP;
+	}
+
+	tx_buf = (drv_data->pins_cfg.outputs_state & ~mask);
+	tx_buf |= (value & mask);
+	tx_buf ^= toggle;
+	tx_buf_p[0] = (uint8_t)tx_buf; // for P07-P00
+	tx_buf_p[1] = tx_buf >> 8;  // for P17-P10
+	rc = i2c_write_dt(&drv_cfg->i2c, tx_buf_p, sizeof(tx_buf_p)); // send P0 values first and P1 values second
+	if (rc != 0) {
```
Comment (ahmed-mrbeam) [inline drivers/gpio/gpio_pcf8575.c]:
```suggestion
	if (rc) {
```

**Review by ahmed-mrbeam — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_pcf8575.c (hunk 3 above)
Comment (amrithvenkat) [reply drivers/gpio/gpio_pcf8575.c:75]: #define	EIO 5		/* I/O error */ from errno.h

File: drivers/gpio/gpio_pcf8575.c (hunk 1 above)
Comment (amrithvenkat) [reply drivers/gpio/gpio_pcf8575.c:60]:
```
if(dev==NULL)
	{
		return -ENODEV;
	}
```
`#define ENODEV 19       /**< No such device */`
would this be better way?

File: drivers/gpio/gpio_pcf8575.c (hunk 7 above)
Comment (amrithvenkat) [reply drivers/gpio/gpio_pcf8575.c]: done

File: drivers/gpio/gpio_pcf8575.c (hunk 4 above)
Comment (amrithvenkat) [reply drivers/gpio/gpio_pcf8575.c]: done

File: drivers/gpio/gpio_pcf8575.c (hunk 1 above)
Comment (ahmed-mrbeam) [reply drivers/gpio/gpio_pcf8575.c:60]: what are the other available error codes ?

Comment (amrithvenkat) [reply drivers/gpio/gpio_pcf8575.c:60]: can be found under zephyr/lib/libc/armstdc/include/errno.h

Comment (amrithvenkat) [reply drivers/gpio/gpio_pcf8575.c:60]: the check for dev==NULL is done already in the calling function. Should it be done again here?

File: drivers/gpio/gpio_pcf8575.c (hunk 6 above)
Comment (amrithvenkat) [reply drivers/gpio/gpio_pcf8575.c:91]: added a check

File: drivers/gpio/gpio_pcf8575.c (hunk 5 above)
Comment (amrithvenkat) [reply drivers/gpio/gpio_pcf8575.c]: implemented the check

**Review by ahmed-mrbeam — APPROVED** [review]

**Pull request merged**